### PR TITLE
Ignore non-voters in leadership transfer

### PR DESCRIPTION
**Pull request by f3nry**
A leader should not initiate an election on a non-voter, causing it to temporarily become a leader for a short period of time.

### CODE REVIEW

File: raft_test.go [hunk 1]
```diff
@@ -2094,6 +2094,29 @@ func TestRaft_LeadershipTransferLeaderReplicationTimeout(t *testing.T) {
 	}
 }
 
+func TestRaft_LeadershipTransferIgnoresNonvoters(t *testing.T) {
+	c := MakeCluster(2, t, nil)
+	defer c.Close()
+
+	follower := c.Followers()[0]
+
+	demoteFuture := c.Leader().DemoteVoter(follower.localID, 0, 1*time.Nanosecond)
```
Comment (briankassouf) [inline raft_test.go:2103]: Could we make this a larger timeout or no timeout? Seems like 1 nanosecond could cause this to be flakey if go doesn't schedule the receiver right away.

Comment (f3nry) [reply raft_test.go]: Thanks for the feedback - I changed this to zero.

**Comment by f3nry** [comment]
@briankassouf Thank you for the review! I'm separately working on signing the CLA.

I'm wondering if there's additional places where we'd want to protect against this occurring or if this is sufficient. For example, should a node to reject a TimeoutNow request if it is currently a non-voter? Should it be an invalid state transition to become a candidate if a node doesn't have a vote?

**Comment by i0rek** [comment]
@f3nry I think you are correct, a non-voter should not become a leader and should reject leadership transfers. 

There is an issue in Consul for your PR: https://github.com/hashicorp/consul/issues/7031. And another one that is similar: https://github.com/hashicorp/consul/issues/6979. The first one should be fixed by this PR. The last one is still open.

**Comment by f3nry** [comment]
@i0rek It seems like the [second bug](https://github.com/hashicorp/raft/blob/72a6ed385dac0c8ee1628d12a9f0bb99abf39f19/raft.go#L216) is mostly protected by the configuration index check on line 207.. Under what conditions would `r.configurations.latestIndex == r.configurations.committedIndex` not pass?

**Comment by f3nry** [comment]
@briankassouf CLA is signed and PR is ready to go!

**Comment by briankassouf** [comment]
@f3nry Great! Thanks!

**Pull request merged**